src/Microsoft.OData.Client/RequestInfo.cs

```diff
@@ -469,9 +469,15 @@ internal DataServiceClientRequestMessage CreateRequestMessage(BuildingRequestEve
             }
             else
             {
-                clientRequestMessage = new HttpWebRequestMessage(clientRequestMessageArgs);
+                if (this.Context.HttpRequestTransportMode == HttpRequestTransportMode.HttpWebRequestMessage)
+                {
+                    clientRequestMessage = new HttpWebRequestMessage(clientRequestMessageArgs);
+                }
+                else 
+                {
+                    clientRequestMessage = new HttpClientRequestMessage(clientRequestMessageArgs);
+                }           

```

Comment:
Maybe this is overkill, by I think we could have a factory class or function. Something like `Func<ClientRequestMessageArgs, IODataRequestMessage>` or a factory class with a `IODataRequestMessage CreateRequestMessage(clientRequestMessageArgs)` method.

Then we add that factory function as property of the `DataServiceContext` based on the `HttpRequestTransportMode`. And then here we just call:

```c#
clientRequestMessage = this.Context.RequestMessageFactory.CreateRequestMessage(clientRequestMessageArgs)
```
And we get rid of the `if` statements. I think this would make it easier to add or remove support for different request message implementations without modifying this class.

Reply:
resolved